Cannot use after delay with user-defined types

If you use a typedef to define type and import a file with the type defined in a file-scoped preamble, the type is not made available to the generated delay reactors.  This makes it very difficult (impossible?) to use after delays with user-defined types unless everything is one file.

To reproduce, put the following into a file `GenDelay.lf`:

```
target C

preamble {=
  #ifndef GEN_DELAY
  #define GEN_DELAY
  typedef int message_t;
  #endif
=}

reactor Source {
  output out: message_t
  reaction(startup) -> out {=
    lf_set(out, 42);
  =}
}

reactor Sink {
  input in: message_t
  reaction(in) {=
    lf_print("Received %d at time %lld", in->value, lf_time_logical_elapsed());
  =}
}
```

Then put this into another file, e.g. `GenDelayTest.lf`:

```
target C
import Source, Sink from "GenDelay.lf"

main reactor {
  source = new Source()
  sink = new Sink()
  source.out -> sink.in after 10 ms
}
```

The file fails to compile with this error:

```
llij/test/C/src-gen/GenDelayTest/./include/GenDelayTest/_lf_GenDelay_49429363.h:39:5: error: unknown type name 'message_t'
    message_t value;
    ^
/Users/eal/lingua-franca-intellij/test/C/src-gen/GenDelayTest/./include/GenDelayTest/_lf_GenDelay_49429363.h:55:5: error: unknown type name 'message_t'
    message_t value;
    ^
In file included from /Users/eal/lingua-franca-intellij/test/C/src-gen/GenDelayTest/__lf_gendelay_49429363.c:3:
/Users/eal/lingua-franca-intellij/test/C/src-gen/GenDelayTest/./__lf_gendelay_49429363.h:13:5: error: unknown type name 'message_t'
    message_t value;
```

--- a/core/src/main/java/org/lflang/ast/DelayedConnectionTransformation.java
+++ b/core/src/main/java/org/lflang/ast/DelayedConnectionTransformation.java
@@ -104,6 +104,7 @@ public class DelayedConnectionTransformation implements AstTransformation {
           Type type = ((Port) connection.getRightPorts().get(0).getVariable()).getType();
           var resource =
               connection.getLeftPorts().size() > 0
+                      && connection.getLeftPorts().get(0).getContainer() != null
                   ? connection.getLeftPorts().get(0).getContainer().getReactorClass().eResource()
                   : mainResource;
           Reactor delayClass = getDelayClass(resource, type, connection.isPhysical());

--- a/core/src/main/java/org/lflang/ast/DelayedConnectionTransformation.java
+++ b/core/src/main/java/org/lflang/ast/DelayedConnectionTransformation.java
@@ -102,7 +102,11 @@ public class DelayedConnectionTransformation implements AstTransformation {
           EObject parent = connection.eContainer();
           // Assume all the types are the same, so just use the first on the right.
           Type type = ((Port) connection.getRightPorts().get(0).getVariable()).getType();
-          Reactor delayClass = getDelayClass(type, connection.isPhysical());
+          var resource =
+              connection.getLeftPorts().size() > 0
+                  ? connection.getLeftPorts().get(0).getContainer().getReactorClass().eResource()
+                  : mainResource;
+          Reactor delayClass = getDelayClass(resource, type, connection.isPhysical());
           String generic = targetTypes.supportsGenerics() ? targetTypes.getTargetType(type) : null;
 
           Instantiation delayInstance =
@@ -274,7 +278,7 @@ public class DelayedConnectionTransformation implements AstTransformation {
    * @param type The type the delay class must be compatible with.
    * @param isPhysical Is this delay reactor using a physical action.
    */
-  private Reactor getDelayClass(Type type, boolean isPhysical) {
+  private Reactor getDelayClass(Resource resource, Type type, boolean isPhysical) {
     String className;
     if (targetTypes.supportsGenerics()) {
       className = DelayBodyGenerator.GEN_DELAY_CLASS_NAME;
@@ -381,7 +385,7 @@ public class DelayedConnectionTransformation implements AstTransformation {
     delayClass.getInputs().add(input);
     delayClass.getOutputs().add(output);
     delayClass.getParameters().add(delayParameter);
-    addDelayClass(delayClass);
+    addDelayClass(resource, delayClass);
     return delayClass;
   }
 
@@ -389,12 +393,11 @@ public class DelayedConnectionTransformation implements AstTransformation {
    * Store the given reactor in the collection of generated delay classes and insert it in the AST
    * under the top-level reactor's node.
    */
-  private void addDelayClass(Reactor generatedDelay) {
+  private void addDelayClass(Resource resource, Reactor generatedDelay) {
     // Record this class, so it can be reused.
     delayClasses.add(generatedDelay);
     // And hook it into the AST.
-    EObject node =
-        IteratorExtensions.findFirst(mainResource.getAllContents(), Model.class::isInstance);
+    EObject node = IteratorExtensions.findFirst(resource.getAllContents(), Model.class::isInstance);
     ((Model) node).getReactors().add(generatedDelay);
   }
 

--- a/test/C/src/GenDelayTest.lf
+++ b/test/C/src/GenDelayTest.lf
@@ -1,9 +1,9 @@
 /**
- * Test that types used in delayed connections do not need to be in scope for
- * the parent reactor in order to compile.
+ * Test that types used in delayed connections do not need to be in scope for the parent reactor in
+ * order to compile.
  */
-
 target C
+
 import Source, Sink from "lib/GenDelay.lf"
 
 main reactor {

--- a/test/C/src/GenDelayTest.lf
+++ b/test/C/src/GenDelayTest.lf
@@ -1,0 +1,13 @@
+/**
+ * Test that types used in delayed connections do not need to be in scope for
+ * the parent reactor in order to compile.
+ */
+
+target C
+import Source, Sink from "lib/GenDelay.lf"
+
+main reactor {
+  source = new Source()
+  sink = new Sink()
+  source.out -> sink.in after 10 ms
+}

--- a/test/C/src/lib/GenDelay.lf
+++ b/test/C/src/lib/GenDelay.lf
@@ -1,0 +1,19 @@
+target C
+
+preamble {=
+  typedef int message_t;
+=}
+
+reactor Source {
+  output out: message_t
+  reaction(startup) -> out {=
+    lf_set(out, 42);
+  =}
+}
+
+reactor Sink {
+  input in: message_t
+  reaction(in) {=
+    lf_print("Received %d at time %lld", in->value, lf_time_logical_elapsed());
+  =}
+}

--- a/test/C/src/lib/GenDelay.lf
+++ b/test/C/src/lib/GenDelay.lf
@@ -1,19 +1,15 @@
 target C
 
-preamble {=
-  typedef int message_t;
-=}
+preamble {= typedef int message_t; =}
 
 reactor Source {
   output out: message_t
-  reaction(startup) -> out {=
-    lf_set(out, 42);
-  =}
+
+  reaction(startup) -> out {= lf_set(out, 42); =}
 }
 
 reactor Sink {
   input in: message_t
-  reaction(in) {=
-    lf_print("Received %d at time %lld", in->value, lf_time_logical_elapsed());
-  =}
+
+  reaction(in) {= lf_print("Received %d at time %lld", in->value, lf_time_logical_elapsed()); =}
 }

